### PR TITLE
Fixing Enqueued Job Trigger for Multiple Queues

### PR DESCRIPTION
**Pull request by Tinyakov**
There was a significant issue with the new enqueued job trigger that should release waiting in jobs fetching loops. It used only one static `AutoResetEvent` instance for all workers and queues. When this `AutoResetEvent` is released, it only releases one random waiting thread. If it was a thread of a worker with a different queue, the job gets stuck until a new event or the `QueuePollInterval` passes. This situation frequently occurred in my environment.

- Introduced new class `AutoResetEventRegistry` in place of one `AutoResetEvent`. It internally holds several `AutoResetEvent` instances, one for each queue.
- Moved its triggering from `PostgreSqlStorage `to `PostgreSqlWriteOnlyTransaction`, where new jobs are added and committed. The list `_queuesWithAddedJobs` accumulates all queues with added jobs in the transaction, and only those queues fire events after commit. This approach also reduces false positive triggers, as not every transaction contains new jobs.
- Minor: Removed `cancellationToken.ThrowIfCancellationRequested()` after waiting in the fetch loop since it is already executed at the beginning of each iteration.

Concerns:
- I didn't make any changes to the `Dequeue_UpdateCount` method, which is used in case of `UseNativeDatabaseTransactions = false`, because there was originally no signal waiting. I'm not sure if it should be modified.

### CODE REVIEW

**Comment by dmitry-vychikov** [comment]
Ideally, this could use postgres notifications channel. https://www.postgresql.org/docs/current/sql-notify.html
Then it will work even for multi process / multi server scenario.

I think there were usages of such approach already in the code.

**Comment by dmitry-vychikov** [comment]
@Tinyakov 

> long-living session/connection which comes at a cost. Hangfire.Postgres provides an option for using it, but I prefer not to enable it.

Do you have any benchmarks or experience to share? How high is that performance cost? Just interested.

I'm not totally against this improvement. But to me using postgres channels looks simpler because it can cover both cases. It will be less code in total to maintain which is a good thing in general. Now it is a bit messy because of combination of multiple approaches in different places.

File: src/Hangfire.PostgreSql/Utils/AutoResetEventRegistry.cs [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+      {
+        if (_events.TryGetValue(eventKey, out AutoResetEvent handle))
+        {
+          yield return handle;
+        }
+        else
+        {
+          AutoResetEvent newHandle = _events.GetOrAdd(eventKey, new AutoResetEvent(false));
+          yield return newHandle;
+        }
```
Comment (dmitry-vychikov) [inline src/Hangfire.PostgreSql/Utils/AutoResetEventRegistry.cs:31]: Why need the `if` statement here? It should work the same if you keep only `else` branch.

```suggestion
      {
          AutoResetEvent handle = _events.GetOrAdd(eventKey, new AutoResetEvent(false));
          yield return handle;
       }
```

Comment (dmitry-vychikov) [reply src/Hangfire.PostgreSql/Utils/AutoResetEventRegistry.cs:31]: You should be using overload with `Func` parameter.

```suggestion
      {
          AutoResetEvent newHandle = _events.GetOrAdd(eventKey, _ => new AutoResetEvent(false));
          yield return newHandle;
```

**Comment by azygis** [comment]
I did leave a few remarks which don't seem to have been addressed yet.

**Comment by azygis** [comment]
Shoot, sorry, I always forget that GitHub requires _submitting_ the review as opposed to Azure DevOps where comments are just appearing automatically. Published it now.

File: src/Hangfire.PostgreSql/PostgreSqlWriteOnlyTransaction.cs [hunk 2]
```diff
@@ -37,6 +37,7 @@ public class PostgreSqlWriteOnlyTransaction : JobStorageTransaction
   {
     private readonly Queue<Action<IDbConnection>> _commandQueue = new();
     private readonly Func<DbConnection> _dedicatedConnectionFunc;
+    private readonly List<string> _queuesWithAddedJobs = new();
```
[the remark this answers is not in the document]
Comment (azygis) [reply src/Hangfire.PostgreSql/PostgreSqlWriteOnlyTransaction.cs:40]: Fair enough, let's see how it goes for now.

**Review by azygis — APPROVED** [review]

**Comment by azygis** [comment]
The package has been published.

**Pull request merged**